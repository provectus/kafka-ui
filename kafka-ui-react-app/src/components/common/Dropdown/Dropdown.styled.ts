import styled, { css, keyframes } from 'styled-components';
import { ControlledMenu } from '@szhsin/react-menu';
import { menuSelector, menuItemSelector } from '@szhsin/react-menu/style-utils';

import '@szhsin/react-menu/dist/core.css';

const menuShow = keyframes`
  from {
    opacity: 0;
  }
`;
const menuHide = keyframes`
  to {
    opacity: 0;
  }
`;

export const Dropdown = styled(ControlledMenu)(
  ({ theme: { dropdown } }) => css`
    // container for the menu items
    ${menuSelector.name} {
      border: 1px solid ${dropdown.borderColor};
      box-shadow: 0 4px 16px ${dropdown.shadow};
      padding: 8px 0;
      border-radius: 4px;
      font-size: 14px;
      background-color: ${dropdown.backgroundColor};
      text-align: left;
    }

    ${menuSelector.stateOpening} {
      animation: ${menuShow} 0.15s ease-out;
    }

    // NOTE: animation-fill-mode: forwards is required to
    // prevent flickering with React 18 createRoot()
    ${menuSelector.stateClosing} {
      animation: ${menuHide} 0.2s ease-out forwards;
    }

    ${menuItemSelector.name} {
      padding: 6px 16px;
      min-width: 150px;
      background-color: ${dropdown.item.backgroundColor.default};
      white-space: nowrap;
    }

    ${menuItemSelector.hover} {
      background-color: ${dropdown.item.backgroundColor.hover};
    }

    ${menuItemSelector.disabled} {
      cursor: not-allowed;
      opacity: 0.5;
    }
  `
);

export const DropdownButton = styled.button`
  background-color: transparent;
  border: none;
  display: flex;
  cursor: pointer;
  align-self: center;

  &:disabled {
    opacity: 0.5;
    cursor: not-allowed;
  }
`;

export const DangerItem = styled.div`
  color: ${({ theme: { dropdown } }) => dropdown.item.color.normal};
`;

export const DropdownItemHint = styled.div`
  color: ${({ theme }) => theme.topicMetaData.color.label};
  font-size: 12px;
  line-height: 1.4;
  margin-top: 5px;
`;

export const Wrapper = styled.div`
  display: inline-flex;
  align-items: center;
  justify-content: end;
  color: ${({ theme: { dropdown } }) => dropdown.item.color.normal};
`;
