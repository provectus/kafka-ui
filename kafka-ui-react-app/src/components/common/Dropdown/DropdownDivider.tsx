import React from 'react';

const DropdownDivider: React.FC = () => <hr className="dropdown-divider" />;

export default DropdownDivider;
