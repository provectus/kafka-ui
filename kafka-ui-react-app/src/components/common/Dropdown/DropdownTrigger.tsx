import { styled } from 'lib/themedStyles';
import React from 'react';

interface Props {
  className?: string;
  onClick: () => void;
}

const DropdownTrigger: React.FC<Props> = ({ onClick, className, children }) => {
  return (
    <button
      className={className}
      type="button"
      aria-haspopup="true"
      aria-controls="dropdown-menu"
      onClick={onClick}
    >
      {children}
    </button>
  );
};

export default styled(DropdownTrigger)`
  background: transparent;
  border: none;
  display: flex;
  align-items: 'center';
  justify-content: 'center';
  &:hover {
    cursor: pointer;
  }
`;
