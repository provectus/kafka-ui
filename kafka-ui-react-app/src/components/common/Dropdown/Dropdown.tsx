import { MenuProps } from '@szhsin/react-menu';
import React, { PropsWithChildren, useRef } from 'react';
import VerticalElipsisIcon from 'components/common/Icons/VerticalElipsisIcon';
import useBoolean from 'lib/hooks/useBoolean';

import * as S from './Dropdown.styled';

interface DropdownProps extends PropsWithChildren<Partial<MenuProps>> {
  label?: React.ReactNode;
  disabled?: boolean;
}

const Dropdown: React.FC<DropdownProps> = ({ label, disabled, children }) => {
  const ref = useRef(null);
  const { value: isOpen, setFalse, setTrue } = useBoolean(false);

  const handleClick: React.MouseEventHandler<HTMLButtonElement> = (e) => {
    e.preventDefault();
    e.stopPropagation();
    setTrue();
  };

  return (
    <S.Wrapper>
      <S.DropdownButton
        onClick={handleClick}
        ref={ref}
        aria-label="Dropdown Toggle"
        disabled={disabled}
      >
        {label || <VerticalElipsisIcon />}
      </S.DropdownButton>
      <S.Dropdown
        anchorRef={ref}
        state={isOpen ? 'open' : 'closed'}
        onMouseLeave={setFalse}
        onClose={setFalse}
        align="end"
        direction="bottom"
        offsetY={10}
        viewScroll="auto"
      >
        {children}
      </S.Dropdown>
    </S.Wrapper>
  );
};

export default Dropdown;
