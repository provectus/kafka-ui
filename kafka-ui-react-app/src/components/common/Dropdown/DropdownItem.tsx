import React, { PropsWithChildren } from 'react';
import { ClickEvent, MenuItem, MenuItemProps } from '@szhsin/react-menu';
import { useConfirm } from 'lib/hooks/useConfirm';

import * as S from './Dropdown.styled';

export interface DropdownItemProps extends PropsWithChildren<MenuItemProps> {
  danger?: boolean;
  onClick?(): void;
  confirm?: React.ReactNode;
}

const DropdownItem = React.forwardRef<unknown, DropdownItemProps>(
  ({ onClick, danger, children, confirm, ...rest }, ref) => {
    const confirmation = useConfirm();

    const handleClick = (e: ClickEvent) => {
      if (!onClick) return;

      // eslint-disable-next-line no-param-reassign
      e.stopPropagation = true;
      e.syntheticEvent.stopPropagation();

      if (confirm) {
        confirmation(confirm, onClick);
      } else {
        onClick();
      }
    };

    return (
      <MenuItem onClick={handleClick} {...rest} ref={ref}>
        {danger ? <S.DangerItem>{children}</S.DangerItem> : children}
      </MenuItem>
    );
  }
);

export default DropdownItem;
