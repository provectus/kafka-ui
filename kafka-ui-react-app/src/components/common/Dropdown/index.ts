import { DropdownItemHint } from './Dropdown.styled';
import Dropdown from './Dropdown';
import DropdownItem from './DropdownItem';

export { Dropdown, DropdownItem, DropdownItemHint };
