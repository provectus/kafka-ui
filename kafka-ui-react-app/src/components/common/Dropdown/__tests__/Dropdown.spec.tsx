import React from 'react';
import { mount } from 'enzyme';
import Dropdown, { DropdownProps } from 'components/common/Dropdown/Dropdown';
import DropdownItem from 'components/common/Dropdown/DropdownItem';
import DropdownDivider from 'components/common/Dropdown/DropdownDivider';

const dummyLable = 'My Test Label';
const dummyChildren = (
  <>
    <DropdownItem onClick={jest.fn()}>Child 1</DropdownItem>
    <DropdownItem onClick={jest.fn()}>Child 2</DropdownItem>
    <DropdownDivider />
    <DropdownItem onClick={jest.fn()}>Child 3</DropdownItem>
  </>
);

describe('Dropdown', () => {
  const setupWrapper = (
    props: Partial<DropdownProps> = {},
    children: React.ReactNode = undefined
  ) => (
    <Dropdown label={dummyLable} {...props}>
      {children}
    </Dropdown>
  );

  it('renders Dropdown with initial props', () => {
    const wrapper = mount(setupWrapper());
    expect(wrapper.exists('.dropdown')).toBeTruthy();

    expect(wrapper.exists('.dropdown.is-active')).toBeFalsy();
    expect(wrapper.exists('.dropdown.is-right')).toBeFalsy();
    expect(wrapper.exists('.dropdown.is-up')).toBeFalsy();

    expect(wrapper.exists('.dropdown-trigger')).toBeTruthy();
    expect(wrapper.find('.dropdown-trigger').text()).toEqual(dummyLable);
    expect(wrapper.exists('.dropdown-content')).toBeTruthy();
    expect(wrapper.find('.dropdown-content').text()).toEqual('');
  });

  it('renders custom children', () => {
    const wrapper = mount(setupWrapper({}, dummyChildren));
    expect(wrapper.exists('.dropdown-content')).toBeTruthy();
    expect(wrapper.find('.dropdown-item').length).toEqual(3);
    expect(wrapper.find('.dropdown-divider').length).toEqual(1);
  });

  it('renders dropdown with a right-aligned menu', () => {
    const wrapper = mount(setupWrapper({ right: true }));
    expect(wrapper.exists('.dropdown.is-right')).toBeTruthy();
  });

  it('renders dropdown with a popup menu', () => {
    const wrapper = mount(setupWrapper({ up: true }));
    expect(wrapper.exists('.dropdown.is-up')).toBeTruthy();
  });

  it('handles click', () => {
    const wrapper = mount(setupWrapper());
    expect(wrapper.exists('button')).toBeTruthy();
    expect(wrapper.exists('.dropdown.is-active')).toBeFalsy();

    wrapper.find('button').simulate('click');
    expect(wrapper.exists('.dropdown.is-active')).toBeTruthy();
  });

  it('matches snapshot', () => {
    const wrapper = mount(
      setupWrapper(
        {
          right: true,
          up: true,
        },
        dummyChildren
      )
    );
    expect(wrapper).toMatchSnapshot();
  });
});
