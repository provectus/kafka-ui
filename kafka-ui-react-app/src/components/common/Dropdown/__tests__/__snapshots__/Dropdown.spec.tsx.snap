// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown matches snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.c1 {
  background: transparent;
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: 'center';
  -webkit-box-align: 'center';
  -ms-flex-align: 'center';
  align-items: 'center';
  -webkit-box-pack: 'center';
  -webkit-justify-content: 'center';
  -ms-flex-pack: 'center';
  justify-content: 'center';
}

.c1:hover {
  cursor: pointer;
}

.c2 {
  color: initial;
}

<body>
  <div>
    <div
      class="dropdown is-right is-up"
    >
      <div
        class="c0"
      >
        <button
          class="c1"
          type="button"
        >
          My Test Label
        </button>
      </div>
      <div
        class="dropdown-menu"
        id="dropdown-menu"
        role="menu"
      >
        <div
          class="dropdown-content has-text-left"
        >
          <a
            class="c2 dropdown-item is-link"
            href="#end"
            role="menuitem"
            type="button"
          >
            Child 1
          </a>
          <a
            class="c2 dropdown-item is-link"
            href="#end"
            role="menuitem"
            type="button"
          >
            Child 2
          </a>
          <hr
            class="dropdown-divider"
          />
          <a
            class="c2 dropdown-item is-link"
            href="#end"
            role="menuitem"
            type="button"
          >
            Child 3
          </a>
        </div>
      </div>
    </div>
  </div>
</body>
`;
